packages/issuer-token-endpoint/lib/IssuerTokenEndpoint.ts

```diff
@@ -0,0 +1,134 @@
+    header: { typ: Typ.JWT, alg: Alg.ES256 },
+    payload: { iat: issuanceTime.getTime(), exp: +new Date(issuanceTime.getTime() + tokenExpiresIn) },
+  }
+  return await jwtSignerCallback(jwt)
+}
+
+const handleTokenRequest = async (request: Request, response: Response) => {
+  response.set({
+    'Cache-Control': 'no-store',
+    Pragma: 'no-cache',
+  })
+
+  const cNonce = v4()
+  await nonceStateManager.setState(cNonce, { cNonce, createdOn: +new Date() })
+  setTimeout(() => {
+    nonceStateManager.deleteState(cNonce)
+  }, cNonceExpiresIn)
+
+  const access_token = await generateAccessToken()
+  const responseBody: AccessTokenResponse = {
+    access_token,
+    token_type: 'bearer',
+    expires_in: tokenExpiresIn,
+    c_nonce: cNonce,
+    c_nonce_expires_in: cNonceExpiresIn,
+    authorization_pending: false,
+    interval,
+  }
+  return response.status(200).json(responseBody)
+}
+
+export const handleHTTPStatus400 = async (request: Request, response: Response, next: NextFunction) => {
+  // TODO invalid_client: the client tried to send a Token Request with a Pre-Authorized Code without Client ID but the Authorization Server does not support anonymous access
+  const assertedState = (await stateManager.getAssertedState(request.body.state)) as CredentialOfferState
+  await stateManager.setState(request.body.state, assertedState)
+  if (request.body.grant_type === 'urn:ietf:params:oauth:grant-type:pre-authorized_code') {
+    if (!request.body['pre-authorized_code']) {
+      return response.status(400).json({ error: 'invalid_request', error_description: 'pre-authorized_code is required' })
+    }
+    /*
+    invalid_request:
+    the Authorization Server expects a PIN in the pre-authorized flow but the client does not provide a PIN
+     */
+    if (assertedState.userPinRequired && !request.body.user_pin) {
+      return response.status(400).json({ error: 'invalid_request', error_description: 'User pin is required' })
+    }
+    /*
+    invalid_request:
+    the Authorization Server does not expect a PIN in the pre-authorized flow but the client provides a PIN
+     */
+    if (!assertedState.userPinRequired && request.body.user_pin) {
+      return response.status(400).json({ error: 'invalid_request', error_description: 'User pin is not required' })
+    }
+    /*
+    invalid_grant:
+    the Authorization Server expects a PIN in the pre-authorized flow but the client provides the wrong PIN
+    the End-User provides the wrong Pre-Authorized Code or the Pre-Authorized Code has expired
+     */
+    if ((request.body.user_pin && !/[0-9{,8}]/.test(request.body.user_pin)) || assertedState.userPin != getNumberOrUndefined(request.body.user_pin)) {
+      return response.status(400).json({ error: 'invalid_grant', error_message: 'PIN must consist of maximum 8 numeric characters' })
+    }
+    const now = +new Date()
+    const expirationTime = assertedState.preAuthorizedCodeCreatedOn + assertedState.preAuthorizedCodeExpiresIn
+    if (
```

Comment:
See remarks about storing these in the state itself (most of them are already there).

I don't like this expression and the error message. These are completely seperate concerns, so why do this in one expression.

Reply:
Fixed!